pkg/cmd/dev/builder.go

```diff
@@ -124,7 +124,26 @@ func (d *dev) getDockerRunArgs(
 	} else {
 		args = append(args, "-i")
 	}
-	args = append(args, "-v", workspace+":/cockroach")
+	gitDir, err := d.exec.CommandContextSilent(ctx, "git", "rev-parse", "--git-dir")
+	if err != nil {
+		return nil, err
+	}
+	gitCommonDir, err := d.exec.CommandContextSilent(ctx, "git", "rev-parse", "--git-common-dir")
+	if err != nil {
+		return nil, err
+	}
+	// If run from inside a git worktree
+	if string(gitDir) != string(gitCommonDir) {
+		mainRepo, err := d.exec.CommandContextSilent(ctx, "cat", ".git")
+		if err != nil {
+			return nil, err
+		}
+		mountPath := strings.Split(strings.Split(string(mainRepo), ": ")[1], "/.git/")[0]
+		args = append(args, "-v", mountPath+":/cockroach")
```

Comment:
Looks backwards: ALWAYS, we want `/cockroach` to be the workspace (the `cockroach` checkout you're running `dev` from, which might be a worktree).